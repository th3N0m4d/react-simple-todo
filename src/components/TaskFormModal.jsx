import React from 'react'
import Button from 'react-bootstrap/Button'
import PropTypes from 'prop-types'
import Modal from 'react-bootstrap/Modal'
import Form from 'react-bootstrap/Form'
import * as R from 'ramda'

import variants from '../constants/Variants'

const defaultTask = {
  name: '',
  author: '',
  variant: variants.info,
  completed: false
}

const propTypes = {
  id: PropTypes.string,
  completed: PropTypes.bool
}

const defaultProps = {
  id: '0',
  completed: false,
  task: defaultTask
}

const mapVariantsToOption = value => <option value={value} key={value}>{value}</option>

const TaskFormModal = ({ task, onHide, onSave, labels, show }) => {
  const [formFields, setFormField] = React.useState(task)

  const handleOnChange = e => {
    setFormField(R.mergeRight(
      formFields,
      { [e.target.name]: e.target.value }
    ))
  }

  const handleOnSave = e => {
    onSave(formFields)
    setFormField(defaultTask)
  }

  return (
    <Modal
      size='lg'
      aria-labelledby='contained-modal-title-vcenter'
      centered
      onHide={onHide}
      show={show}
    >
      <Modal.Header closeButton>
        <Modal.Title id='contained-modal-title-vcenter'>
            Create / Edit Task
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Form>
          <Form.Group>
            <Form.Label>Name</Form.Label>
            <Form.Control
              type='text'
              placeholder='Enter task name'
              name='name'
              value={formFields.name}
              onChange={handleOnChange}
            />
          </Form.Group>
          <Form.Group>
            <Form.Label>Author</Form.Label>
            <Form.Control
              type='text'
              placeholder='Authors name'
              name='author'
              value={formFields.author}
              onChange={handleOnChange}
            />
          </Form.Group>

          <Form.Group>
            <Form.Label>Label color</Form.Label>
            <Form.Control as='select' name='variant' value={formFields.variant} onChange={handleOnChange}>
              {
                R.compose(
                  R.values,
                  R.map(mapVariantsToOption)
                )(variants)

              }
            </Form.Control>
          </Form.Group>

          <Form.Group>
            <Form.Check
              type='checkbox'
              label='Completed'
              name='completed'
              value={formFields.completed}
              onChange={handleOnChange}
            />
          </Form.Group>
        </Form>
      </Modal.Body>
      <Modal.Footer>
        <Button variant='outline-secondary' onClick={onHide}>Close</Button>
        <Button onClick={handleOnSave}>Save</Button>
      </Modal.Footer>
    </Modal>
  )
}

TaskFormModal.propTypes = propTypes
TaskFormModal.defaultProps = defaultProps

export default TaskFormModal
