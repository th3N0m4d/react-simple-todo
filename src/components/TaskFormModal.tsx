import * as React from 'react'
import Button from 'react-bootstrap/Button'
import Modal from 'react-bootstrap/Modal'
import Form from 'react-bootstrap/Form'

interface Props {
  dueDate?: number,
  name?: string,
  onDateChange(dueDate: number): void,
  onHide(): void,
  onNameChange(name: string): void,
  show?: boolean
}

const TaskFormModal: React.FunctionComponent<Props> = ({
  dueDate,
  name,
  onDateChange,
  onHide,
  onNameChange,
  show,
}) => {

  const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    onNameChange(e.target.value)
  }

  const handleDateChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    onDateChange(new Date(e.target.value).getTime())
  }

  return (
    <Modal
      size='lg'
      aria-labelledby='contained-modal-title-vcenter'
      centered={true}
      onHide={onHide}
      show={show}
    >
      <Modal.Header closeButton={true}>
        <Modal.Title id='contained-modal-title-vcenter'>
          Create
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <Form>
          <Form.Group>
            <Form.Label>Name</Form.Label>
            <Form.Control
              type='text'
              name='name'
              value={name}
              onChange={handleNameChange}
            />
          </Form.Group>
          <Form.Group>
            <Form.Label>Due date</Form.Label>
            <Form.Control
              type='date'
              name='dueDate'
              value={dueDate}
              onChange={handleDateChange}
            />
          </Form.Group>
        </Form>
      </Modal.Body>
      <Modal.Footer>
        <Button variant='outline-secondary' data-test="close-button" onClick={onHide}>Close</Button>
        <Button>Save</Button>
      </Modal.Footer>
    </Modal>
  )
}


export default TaskFormModal
