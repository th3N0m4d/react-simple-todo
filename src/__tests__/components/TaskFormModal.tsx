import * as React from 'react'
import { shallow } from 'enzyme'

import TaskFormModal from '@/components/TaskFormModal'

describe('TaskFormModal', () => {
    const defaultProps = {
        onDateChange: jest.fn(),
        onHide: jest.fn(),
        onNameChange: jest.fn()
    }

    describe('Props', () => {
        it('should render default component', () => {
            const wrapper = shallow(<TaskFormModal {...defaultProps} />)

            expect(wrapper).toMatchSnapshot()
        });

        it.each`
            propName        | value
            ${'name'}       | ${'Study React'}
            ${'show'}       | ${true}
            ${'show'}       | ${false}
            ${'dueDate'}    | ${546300000000}
            ${'dueDate'}    | ${null}
        `('should render component with prop $propName and value $value',
            ({ propName, value }) => {
                const props = Object.assign({}, defaultProps, { [propName]: value })
                const wrapper = shallow(<TaskFormModal {...props} />)

                expect(wrapper).toMatchSnapshot()
            })
    });

    describe('Interaction', () => {
        it('should trigger onHide callback', () => {
            const wrapper = shallow(<TaskFormModal {...defaultProps} />)

            const closeButton = wrapper.find('[data-test="close-button"]')

            closeButton.simulate('click')

            expect(defaultProps.onHide).toHaveBeenCalled()
        });

        it('should trigger onNameChange callback', () => {
            const wrapper = shallow(<TaskFormModal {...defaultProps} />)

            const input = wrapper.find('[name="name"]')

            input.simulate('change', { target: { value: 'John Doe' } })

            expect(defaultProps.onNameChange).toHaveBeenCalledWith('John Doe')
        });

        it('should trigger onDateChange callback', () => {
            const wrapper = shallow(<TaskFormModal {...defaultProps} />)

            const input = wrapper.find('[name="dueDate"]')

            input.simulate('change', { target: { value: '2020-01-01' } })

            expect(defaultProps.onDateChange).toHaveBeenCalledWith(1577836800000)
        });
    });

});