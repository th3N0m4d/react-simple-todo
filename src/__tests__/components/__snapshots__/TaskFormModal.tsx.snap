// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskFormModal Props should render component with prop dueDate and value 546300000000 1`] = `
<Bootstrap(Modal)
  aria-labelledby="contained-modal-title-vcenter"
  centered={true}
  onHide={[MockFunction]}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      id="contained-modal-title-vcenter"
    >
      Create
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <Form
      inline={false}
    >
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Name
        </FormLabel>
        <FormControl
          name="name"
          onChange={[Function]}
          type="text"
        />
      </FormGroup>
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Due date
        </FormLabel>
        <FormControl
          name="dueDate"
          onChange={[Function]}
          type="date"
          value={546300000000}
        />
      </FormGroup>
    </Form>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      data-test="close-button"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
      variant="outline-secondary"
    >
      Close
    </Button>
    <Button
      active={false}
      disabled={false}
      type="button"
      variant="primary"
    >
      Save
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;

exports[`TaskFormModal Props should render component with prop dueDate and value null 1`] = `
<Bootstrap(Modal)
  aria-labelledby="contained-modal-title-vcenter"
  centered={true}
  onHide={[MockFunction]}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      id="contained-modal-title-vcenter"
    >
      Create
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <Form
      inline={false}
    >
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Name
        </FormLabel>
        <FormControl
          name="name"
          onChange={[Function]}
          type="text"
        />
      </FormGroup>
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Due date
        </FormLabel>
        <FormControl
          name="dueDate"
          onChange={[Function]}
          type="date"
          value={null}
        />
      </FormGroup>
    </Form>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      data-test="close-button"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
      variant="outline-secondary"
    >
      Close
    </Button>
    <Button
      active={false}
      disabled={false}
      type="button"
      variant="primary"
    >
      Save
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;

exports[`TaskFormModal Props should render component with prop name and value Study React 1`] = `
<Bootstrap(Modal)
  aria-labelledby="contained-modal-title-vcenter"
  centered={true}
  onHide={[MockFunction]}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      id="contained-modal-title-vcenter"
    >
      Create
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <Form
      inline={false}
    >
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Name
        </FormLabel>
        <FormControl
          name="name"
          onChange={[Function]}
          type="text"
          value="Study React"
        />
      </FormGroup>
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Due date
        </FormLabel>
        <FormControl
          name="dueDate"
          onChange={[Function]}
          type="date"
        />
      </FormGroup>
    </Form>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      data-test="close-button"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
      variant="outline-secondary"
    >
      Close
    </Button>
    <Button
      active={false}
      disabled={false}
      type="button"
      variant="primary"
    >
      Save
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;

exports[`TaskFormModal Props should render component with prop show and value false 1`] = `
<Bootstrap(Modal)
  aria-labelledby="contained-modal-title-vcenter"
  centered={true}
  onHide={[MockFunction]}
  show={false}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      id="contained-modal-title-vcenter"
    >
      Create
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <Form
      inline={false}
    >
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Name
        </FormLabel>
        <FormControl
          name="name"
          onChange={[Function]}
          type="text"
        />
      </FormGroup>
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Due date
        </FormLabel>
        <FormControl
          name="dueDate"
          onChange={[Function]}
          type="date"
        />
      </FormGroup>
    </Form>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      data-test="close-button"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
      variant="outline-secondary"
    >
      Close
    </Button>
    <Button
      active={false}
      disabled={false}
      type="button"
      variant="primary"
    >
      Save
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;

exports[`TaskFormModal Props should render component with prop show and value true 1`] = `
<Bootstrap(Modal)
  aria-labelledby="contained-modal-title-vcenter"
  centered={true}
  onHide={[MockFunction]}
  show={true}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      id="contained-modal-title-vcenter"
    >
      Create
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <Form
      inline={false}
    >
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Name
        </FormLabel>
        <FormControl
          name="name"
          onChange={[Function]}
          type="text"
        />
      </FormGroup>
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Due date
        </FormLabel>
        <FormControl
          name="dueDate"
          onChange={[Function]}
          type="date"
        />
      </FormGroup>
    </Form>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      data-test="close-button"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
      variant="outline-secondary"
    >
      Close
    </Button>
    <Button
      active={false}
      disabled={false}
      type="button"
      variant="primary"
    >
      Save
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;

exports[`TaskFormModal Props should render default component 1`] = `
<Bootstrap(Modal)
  aria-labelledby="contained-modal-title-vcenter"
  centered={true}
  onHide={[MockFunction]}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      id="contained-modal-title-vcenter"
    >
      Create
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <Form
      inline={false}
    >
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Name
        </FormLabel>
        <FormControl
          name="name"
          onChange={[Function]}
          type="text"
        />
      </FormGroup>
      <FormGroup>
        <FormLabel
          column={false}
          srOnly={false}
        >
          Due date
        </FormLabel>
        <FormControl
          name="dueDate"
          onChange={[Function]}
          type="date"
        />
      </FormGroup>
    </Form>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      data-test="close-button"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
      variant="outline-secondary"
    >
      Close
    </Button>
    <Button
      active={false}
      disabled={false}
      type="button"
      variant="primary"
    >
      Save
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;
